### PR TITLE
Complexity

### PR DESCRIPTION
**Pull request by Daynvheur**
This third PR is (hopefully) the good one, including every features the 2 previous ones featured, and some tweaks.
- Limit complexity resolution
- View/hide complexity circles
- View complexity path (when set; beware, higher values will have an awfull impact on performances)
- Set `parameter` and `complexity` to be `fftSize`d (performance increase!)

To investigate: @jnm2 would you accept to detail this point?
- Complexity path is missing starting loop on low limits (->draw a 4-leaves clover, limit complexity to 2, a triangle-shaped complexity path is drawn)

### CODE REVIEW

File: src/index.html
Comment (jnm2) [inline src/index.html]: Are self-closing tags generally preferred in HTML5? I was under the impression that they weren't.

What do you think about this? https://www.reddit.com/r/web_design/comments/uvvxg/html5_developers_do_you_self_close_your_tags_or/c4z1416/
The thing I find most convincing is the fact that `<span/><span/>` in HTML5 is the same as `<span><span></span></span>`, so I think in fact HTML5 doesn't have a concept of self-closing, making it misleading. But I'm open to learning.

File: src/index.ts
Comment (jnm2) [inline src/index.ts]: I think I want to disagree with Ctrl+K/D here unless you think spaces are more mainstream.

File: src/index.html
Comment (jnm2) [inline src/index.html:1]: I think this means a byte order mark is being added. I'd prefer not to have byte order marks based on the recommendations I've heard. (I'm fixing this by adding an .editorconfig file and linting.)

File: src/index.ts
Comment (jnm2) [inline src/index.ts]: I'm guessing the cast is no longer needed?

**Review by jnm2 — COMMENTED** [review]
I looked at all the peripheral parts first in this batch of comments. I'm working on a https://github.com/jnm2/epic/tree/lint branch that adds linting and I'd like to add CI to prevent PRs from merging if linting doesn't succeed. However, this conflicts with three kinds of formatting changes that you made in this PR, depending on the discussion in the comments: "Self-closing" HTML5 tags, .html file byte order mark, and putting a space between `function` and `()`.

So I'm thinking that it's not fair to introduce extra checks and force you to rebase the PR branch. I can revert some of these changes myself after merging your PR. Or, if you would prefer, either you or I could rebase your PR branch to remove the formatting commits.

Next I'll look at the actual topic of the PR. 😊

File: src/index.html
Comment (jnm2) [reply src/index.html]: https://html-validate.org/rules/void-style.html also defaults to erroring on self-closed void elements.

Comment (Daynvheur) [reply src/index.html]: Pro:
> XHTML-compatible HTML

Cons:
> They carry exactly no meaning.

To me, the self-closing tag do carry a meaning: _this html(5) tag is not missing/hiding a closing tag somewhere_.
Yes, your particular page is simple and well-formatted enough to not have this kind of questioning, but it's a clear (expressed) intent  that does not carry downsides.
I'm not `/` closing self-closing tags for the `x`html compatibility (that I've no care of), but for human readings and page maintenance.

That said, I'm not changing my general habits though I will respect whatever rules you want to apply on your domain.

File: src/index.ts
Comment (Daynvheur) [reply src/index.ts]: I think you're right on this, Ctrl+K/D did some overwork, I'll try to stick `function` to its parenthesis.

File: src/index.html
Comment (Daynvheur) [reply src/index.html:1]: I don't even know what kind of byte order came here, might be `\n\r` becoming `\r\n`, but that was not intended. I'll try with Notepad++, thanks pointing that out.

File: src/index.ts
Comment (Daynvheur) [reply src/index.ts]: Ah, you're right. I'll remove the cast.

File: src/index.html
Comment (Daynvheur) [reply src/index.html:1]: Line feeds are back in 1176e0d.

File: src/index.ts
Comment (Daynvheur) [reply src/index.ts]: Cast removed in c23386f.

File: src/index.html
Comment (Daynvheur) [reply src/index.html]: Html5 self-closing syntax applied everywhere (even on `<meta>` tags) in 1176e0d.

File: src/index.ts
Comment (Daynvheur) [reply src/index.ts]: Removed the useless space in b340b2b.

**Comment by Daynvheur** [comment]
Do apply the changes you want, as you're the repo owner: I'll adapt my fork to be rule-compliant, you won't bother me.

_[Nota on the html5 self-closing syntax]
My opinion stand as expressing it does have a meaning for anyone who reads the tags without prior knowledge on which tags are automagically self-closables (`meta`, `input`), those that are optionally self-closed (`head`, `title`, `div`, `label`, `canvas`), and those that can't be self-closed even when empty (`body`, `style`, `script`).
[/Nota]_

**Comment by jnm2** [comment]
Cool. I'm still forming my thinking, so I appreciate that you are sharing your thoughts.

> My opinion stand as expressing it does have a meaning for anyone who reads the tags without prior knowledge on which tags are automagically self-closables (meta, input), those that are optionally self-closed (head, title, div, label, canvas), and those that can't be self-closed even when empty (body, style, script).

I saw this as a huge benefit until I remembered that the indentation itself clarifies which elements are void elements if the next element starts on a new line. It's very clear by the indentation that meta and input elements don't contain the next element.

My concern on the other side is that self-closing is an XML concept that doesn't work in HTML; the self-closing slash is always ignored in HTML, meaning that the syntax can be misleading (e.g. `<span/><span/>`). Having meta and inputs use a self-closing slash then seems misleading because it seems to indicate that non-void elements can use it too. However, this is only a concern if I extend this style of writing to projects where I'm not using a linter because the linter will call out the problem with `<span/>`.

So I'm back to having no huge benefits or downsides either way. The tiebreaker for me is then that HTML does not have a concept of self-closing. Void tags are not self-closing because they are never open. Non-void tags are not self-closing because HTML doesn't have that concept. I'd rather use the style that shows clearly that the document is HTML, not XML rather than a possibly misleading cross between them. And to keep it readable, I'll use indentation to demonstrate that void tags do not contain the next tag, in case memory fails.

**Comment by jnm2** [comment]
We might want to draw the red line on top of the circles:
image

**Comment by jnm2** [comment]
Is 'complexity circles' a common term? I think we might be able to say 'Show circles' there.

**Comment by jnm2** [comment]
Wow, this is super cool!
image

**Comment by jnm2** [comment]
@Daynvheur I think what's going on with the four-leafed clover is that the curve is not drawn at a constant speed. The parametric curve spends less time in the starting leaf, so it is deprioritized. If we looked at separate graphs of x(t) and y(t) instead of y(x), I think the one-dimensional waveforms would make a lot more sense to us.

The reason I think it's not missing an early epicycle is that it goes on to bring in the faster leaf in when you pile up enough epicycles. Adding another epicycle would cause the end result to go off course when you increase the number of epicycles.
image

At any rate, the non-constant speed and the effect that it has on the epicycles is a preexisting phenomenon that I was noticing before this PR. I don't think that you caused a regression in the calculations. Let me know when you'd like to merge and then I'll see what I can do about the preexisting rate of speed problem afterwards. I'm expecting it will require time smoothing.

Another example:
image

**Comment by jnm2** [comment]
For what it's worth, I can get it to prioritize the 4x speed circle over the 3x speed circle:
image

image

**Comment by Daynvheur** [comment]
I made a comment, didn't pressed `Comment` Button, then switched tab. Comment is gone. :(

Hello @jnm2!
So, in a nutshell:
- 4x speed circles looks cool.
- I agree with the indentation showing intent, but misindentation might hide intent ; my old `/` habits being deprecated with html5, I may be teached with linter force without much effort, I'll be used to not-self-close in a matter of commits. (`<span/>` and `<head/>` are expanded anyway, it's time to drop the self-closing syntax, yay! (That has a measurable impact on page's character load, but, erm, nevermind.)) - 05929f2 got rid of the (introduced) last self-closed tag: `<canvas/>`
- 'Complexity circles' was likely a placeholder name as it displays/hides circles around the complexity rays; `Show circles` is valid, I'll take it. _(In French, we tend to announce what the action targets, instead of what the action merely does (because, you know, 'Show circles' wouldn't let you know what kind of circles will be shown); `Show complexity circles` is likely to be the resulting label.)_  - 05929f2
- `Draw the red line on top of the circle`: are you referring to a fixed-north-like ray? Or to an optimization I could make to generate the circles from the actual destination point? I'll explore both questions anyway. - 104ae44 comments the inefficiency of starting drawing from the `[new_x; new_y]` point; 58f3dd7 moved the starting path from the center, leaving a visual horizontal ray as an immutable reference _(targetting the top-y point would either require one more line between said point and center before moving to rightmost-x, either a `[1.5*Math.PI; 2.5*Math.PI]` angle calculation)_.

I'll send a couple of commits with the late discussion, I'll let you merge when you feel ok after that.

**Comment by Daynvheur** [comment]
The PR is ready to merge.
image

**Comment by jnm2** [comment]
Everything you're doing looks great!

> Draw the red line on top of the circle

With this I meant Z-order (drawing order). When the circles are drawn, it can be hard to see the red lines because the red lines get covered with the circles' color.

**Comment by Daynvheur** [comment]
> I meant Z-order (drawing order).

Arf, I won't be able to change as it would require to invert the drawing order of the arcs before the line path.
Hm, wait a minute, that would be in fact an optimization, thanks.

**Review by jnm2 — APPROVED** [review]
Thanks for these neat features!

**Pull request merged**
